source/common/http/conn_manager_impl.cc

```diff
@@ -197,8 +203,19 @@ StreamDecoder& ConnectionManagerImpl::newStream(StreamEncoder& response_encoder)
   if (read_callbacks_->connection().aboveHighWatermark()) {
     new_stream->callHighWatermarkCallbacks();
   }
+
   new_stream->moveIntoList(std::move(new_stream), streams_);
-  return **streams_.begin();
+
+  ActiveStream* stream = streams_.begin()->get();
+
+  if (overload_stop_accepting_streams_ == Server::OverloadActionState::Active) {
+    read_callbacks_->connection().dispatcher().post([this, stream] {
```

Comment:
Presumably the post here was because doing a direct response from within the newStream callback does not work? (Not surprising). Ultimately, it would be nice if the newStream() callback could do the direct response immediately without a dispatcher post, but it seems fine to indicate that via comment/TODO?

Reply:
Actually, I just tried the unit tests with a direct response and it worked fine so I'll send an update to do that.

Reply:
I'm not completely convinced this will work in the real server but I'm not sure. Can you look into adding an integration test? I think for this type of feature it would be great to have one. I'm thinking you could develop an overload plugin just for integration tests that you can completely control and it probably wouldn't be that hard?

Reply:
Sure, will look into adding an integration test.

Reply:
Added an integration test for this. Also moved this code that drops new requests (by doing a sendLocalReply) to after envoy has decoded the client headers, rather than when we create the stream which is too early to close when using http2.